packages/remix-react/components.tsx

```diff
@@ -45,354 +54,130 @@ import {
-    // deeper in the tree.
-    if (appState.trackCatchBoundaries) {
-      appState.catchBoundaryRouteId = id;
-    }
-
-    context = maybeServerCaught
-      ? {
-          id,
-          get data() {
-            console.error("You cannot `useLoaderData` in a catch boundary.");
-            return undefined;
-          },
-        }
-      : { id, data };
+  let error = useRouteError();
+  let location = useLocation();
+  let { CatchBoundary, ErrorBoundary } = routeModules[id];
 
-    element = (
-      <RemixCatchBoundary
-        location={location}
-        component={CatchBoundary}
-        catch={maybeServerCaught}
-      >
-        {element}
-      </RemixCatchBoundary>
-    );
+  // Provide defaults for the root route if they are not present
+  if (id === "root") {
+    CatchBoundary ||= RemixRootDefaultCatchBoundary;
+    ErrorBoundary ||= RemixRootDefaultErrorBoundary;
   }
 
-  // Only wrap in error boundary if the route defined one, otherwise let the
-  // error bubble to the parent boundary. We could default to using error
-  // boundaries around every route, but now if the app doesn't want users
-  // seeing the default Remix ErrorBoundary component, they *must* define an
-  // error boundary for *every* route and that would be annoying. Might as
-  // well make it required at that point.
-  //
-  // By conditionally wrapping like this, we allow apps to define a top level
-  // ErrorBoundary component and be done with it. Then, if they want to, they
-  // can add more specific boundaries by exporting ErrorBoundary components
-  // for whichever routes they please.
-  //
-  // NOTE: this kind of logic will move into React Router
-
-  if (ErrorBoundary) {
-    // If we tried to render and failed, and this route threw the error, find it
-    // and pass it to the ErrorBoundary to emulate `componentDidCatch`
-    let maybeServerRenderError =
-      appState.error &&
-      (appState.renderBoundaryRouteId === id ||
-        appState.loaderBoundaryRouteId === id)
-        ? deserializeError(appState.error)
-        : undefined;
-
-    // This needs to run after we check for the error from a previous render,
-    // otherwise we will incorrectly render this boundary for a loader error
-    // deeper in the tree.
-    if (appState.trackBoundaries) {
-      appState.renderBoundaryRouteId = id;
+  if (isRouteErrorResponse(error)) {
+    if (error?.error && error.status !== 404) {
+      return (
+        // TODO: Handle error type?
+        <RemixErrorBoundary
+          location={location}
+          component={ErrorBoundary}
+          error={error.error}
+        />
+      );
     }
+    return <RemixCatchBoundary component={CatchBoundary} catch={error} />;
+  }
 
-    context = maybeServerRenderError
-      ? {
-          id,
-          get data() {
-            console.error("You cannot `useLoaderData` in an error boundary.");
-            return undefined;
-          },
```

Comment:
I think this functionality is probably still missing - should get this added to the router

Reply:
Bring this over to router - we will have a larger conversation about this in the future